cluster_client.cpp

```diff
@@ -0,0 +1,358 @@
+        int max_slot = strtol(shard->mbulk_array[1]->value, NULL, 10);
+
+        // hostname/ip
+        mbulk_element* mbulk_addr_el = shard->mbulk_array[2]->mbulk_array[0];
+        char* addr = (char*) malloc(mbulk_addr_el->value_len + 1);
+        memcpy(addr, mbulk_addr_el->value, mbulk_addr_el->value_len);
+        addr[mbulk_addr_el->value_len] = '\0';
+
+        // port
+        mbulk_element* mbulk_port_el = shard->mbulk_array[2]->mbulk_array[1];
+        char* port = (char*) malloc(mbulk_port_el->value_len + 1);
+        memcpy(port, mbulk_port_el->value, mbulk_port_el->value_len);
+        port[mbulk_port_el->value_len] = '\0';
+
+        // check if connection already exist
+        shard_connection* sc = NULL;
+        unsigned int j;
+
+        for (j = 0; j < m_connections.size(); j++) {
+            if (strcmp(addr, m_connections[j]->get_address()) == 0 &&
+                strcmp(port, m_connections[j]->get_port()) == 0) {
+                sc = m_connections[j];
+                break;
+            }
+        }
+
+        // if connection doesn't exist, add it
+        if (sc == NULL) {
+            sc = add_shard_connection(addr, port);
+            assert(sc != NULL);
+        }
+
+        // update range
+        //sc->add_slots_range(min_slot, max_slot);
+        for (int j = min_slot; j <= max_slot; j++) {
+            m_slot_to_shard[j] = sc->get_id();
+        }
+
+        free(addr);
+        free(port);
+    }
+}
+
+bool cluster_client::hold_pipeline(unsigned int conn_id) {
+    // don't exceed requests
+    if (m_config->requests) {
+        if (m_key_index_pools[conn_id]->empty() &&
+            m_reqs_generated >= m_config->requests) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+bool cluster_client::get_key_for_conn(unsigned int conn_id, int iter, unsigned long long* key_index) {
+    // first check if we already have key in pool
+    if (!m_key_index_pools[conn_id]->empty()) {
+        *key_index = m_key_index_pools[conn_id]->front();
+        m_key_len = snprintf(m_key_buffer, sizeof(m_key_buffer)-1, "%s%llu", m_obj_gen->get_key_prefix(), *key_index);
+
+        m_key_index_pools[conn_id]->pop();
+        return true;
+    }
+
+    // keep generate key till it match for this connection, or requests reached
+    while (true) {
+        // generate key
+        *key_index = m_obj_gen->get_key_index(iter);
+        m_key_len = snprintf(m_key_buffer, sizeof(m_key_buffer)-1, "%s%llu", m_obj_gen->get_key_prefix(), *key_index);
+
+        // check if the key match for this connection
+        unsigned int hslot = calc_hslot_crc16_cluster(m_key_buffer, m_key_len);
+        if (m_slot_to_shard[hslot] == conn_id) {
+            m_reqs_generated++;
+            return true;
+        }
+
+        // store key for other connection, if queue is not full
+        key_index_pool* key_idx_pool = m_key_index_pools[m_slot_to_shard[hslot]];
+        if (key_idx_pool->size() < KEY_INDEX_QUEUE_MAX_SIZE) {
```

Comment:
Did you confirm that in all possible scenarios it is ok for the obj_gen to return a key that is simply discarded (because the queue is full)?

Reply:
The answer divide for two parts:
1. For Random/Gaussian:
We think that it is ok for those options, since the discarding of keys suppose to be relative to the selected pattern, so in the overall the distribution will stay as it suppose to be (Random/Gaussian).

2. For Sequential:
Here we have problem that we know about it...
The issue start when the different connections are not symmetrical (networking/shards conuts ...).

In that case one connection can be slow from other and his queue starting to grow over and over and the cluster-client start to discard keys.
In sequential distribution we have issue if user ask for keys (for ex) between  1 - 1000000.

If user set memtier to --test-time and not --requests option, there isn't really issue.
Since the connections are independent, it could be gaps in the DB even if cluster-client will not discard any keys.

The problem is only when user set memtier to --requests option, and if he ask 1000000 request in range of 1 - 1000000 he expect that at the end of the test, the DB will be with those exactly keys (1-10000000), and since we discards keys this will not be the result.

We think about two solutions for this case:
1. add option for user to set the queue limit size.
2. add option to do some balancing between the connections, so if some connection reach to some threshold, the other connection start to send less requests or stop send request, until the slower connection go down below the the threshold.

What do you think?
Is this issue is critical? or we can solved in the future.
The solutions are good, do you have any other idea?

Reply:
I don't think it's critical, it should probably be documented though in a section about cluster support caveats.